### PR TITLE
RE: Adjust the hero API code to deliver uploaded images

### LINKED ISSUE #14473: Adjust the hero API code to deliver uploaded images
Supports https://github.com/mozilla/addons-server/issues/14471
Depends on #14472, #14474 

After #14472 and #14474 land, adjust the code for the hero API to deliver images from the location where the uploaded files are stored, as opposed to from the CDN where they are currently stored.

Question: Will that new location also be served via a CDN, and if not are there any issues with not serving the images from a CDN?

### PR DESCRIPTION
**Pull request by xlisachan**
Fixes #14473

Started a new pull request after yesterday's ordeal.

As mentioned in the previous PR, the image_url property has been updated in the PrimaryHero model (hero/models.py), which is what the serializer was using for the hero image in hero/serializers.py. The following tests have been revised:
- `discovery/tests/test_admin.py`
- `hero/tests/test_models.py`
- `hero/tests/test_serializers.py`

In addition, the updates that were discussed were made as well:
- re-wrote the image_url property in the `return xxx if yyy else zzz` style, adding parenthesis to appease flake8
- added `null=True` to image in PrimaryHero (migration file included)

### CODE REVIEW

**Comment by xlisachan** [comment]
Updates have been made to also pass the following tests:
- `hero/tests/test_models.py`
- `hero/tests/test_views.py`

I wanted to check if the update for `def test_basic` in `hero/tests/test_views.py` was correct. Per the screenshot below, it had initially errored as the number of queries was asserted to be 11, but with the update, there were 3 additional queries for fetching primary hero images.  Therefore, I increased the number of queries to 14. Please let me know if that checks out.

Screen Shot 2020-07-28 at 12 12 53 PM

In addition, as we had discussed regarding the preview of the thumbnails for the primary hero images, I updated the stylesheet so it would not distort the thumbnails.

Please review when you have a moment. Thanks!

**Comment by eviljeff** [comment]
> I wanted to check if the update for def test_basic in hero/tests/test_views.py was correct. Per the screenshot below, it had initially errored as the number of queries was asserted to be 11, but with the update, there were 3 additional queries for fetching primary hero images. Therefore, I increased the number of queries to 14. Please let me know if that checks out.

It means there's an extra query being made for each shelf, which we don't really want.  Try adding 'select_image' to the select_related args in https://github.com/mozilla/addons-server/blob/728925609b4ec2e06f87e67582c9b9b1f49f0daf/src/olympia/hero/views.py#L47

File: src/olympia/hero/tests/test_models.py [hunk 1]
```diff
@@ -68,31 +75,28 @@ def test_clean_gradient_and_image(self):
...
-            gradient_color='#C60184', image='foo.png')
+            gradient_color='#C60184', select_image=self.phi)
         PromotedApproval.objects.create(
             version=hero.promoted_addon.addon.current_version,
             group_id=RECOMMENDED.id)
         assert not hero.enabled
         assert not PrimaryHero.objects.filter(enabled=True).exists()
         # It should still validate even if there are no other enabled shelves,
-        # because we're not changing its enabled state.
-        hero.clean()  # it raises if there's an error
```
Comment (eviljeff) [inline src/olympia/hero/tests/test_models.py:94]: does this fail now?

File: src/olympia/hero/tests/test_views.py [hunk 2]
```diff
@@ -63,6 +78,7 @@ def test_basic(self):
             # - 1 to fetch the add-ons authors
             # - 1 to fetch the add-ons version previews (for static themes)
             # - 1 to fetch the add-ons previews
+            # - 3x: 1 to fetch the primary hero image
```
Comment (eviljeff) [inline src/olympia/hero/tests/test_views.py:81]: See my comment - we should be able to get this back to 11 by using `select_related`

File: src/olympia/discovery/tests/test_admin.py [hunk 3]
```diff
@@ -19,6 +19,22 @@ class TestDiscoveryAdmin(TestCase):
     def setUp(self):
         self.list_url = reverse('admin:discovery_discoveryitem_changelist')
 
+    def _get_heroform(self, item_id):
```
Comment (eviljeff) [inline src/olympia/discovery/tests/test_admin.py:22]: The tests in this file got removed (moved to https://github.com/mozilla/addons-server/blob/master/src/olympia/promoted/tests/test_admin.py but I think I already made the relevant changes when I moved them) so I'm not sure there's anything left you need to change in discover/tests/test_admin.py

File: src/olympia/hero/tests/test_models.py [hunk 4]
```diff
@@ -1,32 +1,35 @@
...
+    def setUp(self):
+        uploaded_photo = get_uploaded_file('transparent.png')
+        self.phi = PrimaryHeroImage.objects.create(custom_image=uploaded_photo)
+
     def test_image_url(self):
         ph = PrimaryHero.objects.create(
-            promoted_addon=PromotedAddon.objects.create(addon=addon_factory()),
-            image='foo.png')
+            disco_addon=DiscoveryItem.objects.create(addon=addon_factory()),
+            select_image=self.phi)
```
Comment (eviljeff) [inline src/olympia/hero/tests/test_models.py:18]: you don't want this change - PrimaryHero doesn't fk to DiscoveryItem any longer.  It's just the kwarg change from `image` that's needed

File: src/olympia/hero/tests/test_models.py [hunk 5]
```diff
@@ -1,32 +1,35 @@
...
             gradient_color='#C60084')
         assert ph.gradient == {'start': 'color-ink-80', 'end': 'color-pink-70'}
 
     def test_clean_requires_recommended(self):
         ph = PrimaryHero.objects.create(
-            promoted_addon=PromotedAddon.objects.create(
-                addon=addon_factory(), group_id=RECOMMENDED.id),
-            gradient_color='#C60184', image='foo.png')
+            disco_addon=DiscoveryItem.objects.create(addon=addon_factory()),
+            gradient_color='#C60184', select_image=self.phi)
```
Comment (eviljeff) [inline src/olympia/hero/tests/test_models.py:32]: same here.

File: src/olympia/hero/tests/test_models.py [hunk 6]
```diff
@@ -42,8 +45,8 @@ def test_clean_requires_recommended(self):
 
     def test_clean_external_requires_homepage(self):
         ph = PrimaryHero.objects.create(
-            promoted_addon=PromotedAddon.objects.create(addon=addon_factory()),
-            is_external=True, gradient_color='#C60184', image='foo.png')
+            disco_addon=DiscoveryItem.objects.create(addon=addon_factory()),
+            is_external=True, gradient_color='#C60184', select_image=self.phi)
```
Comment (eviljeff) [inline src/olympia/hero/tests/test_models.py:49]: and here - the import changes at the top of the file are likely incorrect now too.

File: src/olympia/hero/tests/test_models.py [hunk 7]
```diff
@@ -68,25 +71,24 @@ def test_clean_gradient_and_image(self):
...
-                addon=addon_factory(), group_id=RECOMMENDED.id),
-            gradient_color='#C60184', image='foo.png')
-        PromotedApproval.objects.create(
-            version=hero.promoted_addon.addon.current_version,
-            group_id=RECOMMENDED.id)
+            disco_addon=DiscoveryItem.objects.create(addon=addon_factory()),
+            gradient_color='#C60184', select_image=self.phi)
+        hero.disco_addon.update(recommendable=True)
+        hero.disco_addon.addon.current_version.update(
+            recommendation_approved=True)
```
Comment (eviljeff) [inline src/olympia/hero/tests/test_models.py:91]: and here - recommended status is now derived from PromotedAddon and PromotedApproval

File: src/olympia/hero/tests/test_models.py (hunk 1 above)
Comment (xlisachan) [reply src/olympia/hero/tests/test_models.py]: This was accidentally removed, it has been added back in in the latest push and passes.

File: src/olympia/discovery/tests/test_admin.py (hunk 3 above)
Comment (xlisachan) [reply src/olympia/discovery/tests/test_admin.py]: It has now been removed in the latest push.

**Comment by xlisachan** [comment]
The latest push includes all updates and resolves the merge conflicts. The following Travis CI check fails:
- `GITHUB_TOKEN=[secure]  TOXENV=es ES_VERSION=6.x` with the error `curl: (7) Failed to connect to localhost port 9200: Connection refused`.

I think this related to the error I receive in the other issue when trying to use `settings.EXTERNAL_SITE_URL`. I am still uncertain how to resolve it as `docker-compose ps` states all containers are up.

Please review when you have a moment. Thank you!

**Comment by eviljeff** [comment]
reverted in https://github.com/mozilla/addons-server/pull/15199 because of image size problems we found.  We will re-merge https://github.com/mozilla/addons-server/commit/84c45610be685c69e4f5985a99d667a9626144a8 some time after August 12th (when Scott is back).

**Pull request merged**